File: backend_node/src/services/notificationService.js
import pool from '../config/database.js';
import { sendEmail } from './emailService.js';
import { broadcastToAllUsers } from '../websocket/notificationWebSocket.js';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import dotenv from 'dotenv';

dotenv.config();

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export const sendLivestreamStartNotification = async (livestreamId, livestreamData) => {
  try {
    const users = await pool.query(
      `SELECT id, email, name, notification_preferences FROM users`
    );

    const notifications = users.rows.map(user => ({
      recipient_id: user.id,
      recipient_email: user.email,
      type: 'Livestream Started',
      message: `${livestreamData.title} is now live!`,
      status: 'pending',
      link: '/member-dashboard'
    }));

    if (notifications.length > 0) {
      const values = notifications.map((n, i) => 
        `($${i*6+1}, $${i*6+2}, $${i*6+3}, $${i*6+4}, $${i*6+5}, $${i*6+6})`
      ).join(',');
      
      const params = notifications.flatMap(n => [
        n.recipient_id, n.recipient_email, n.type, n.message, n.status, n.link
      ]);

      await pool.query(
        `INSERT INTO notifications (recipient_id, recipient_email, type, message, status, link) 
         VALUES ${values}`,
        params
      );

      broadcastToAllUsers();

      setImmediate(async () => {
        const watchUrl = `${process.env.FRONTEND_URL}/member-dashboard`;
        
        const templatePath = path.join(__dirname, '../templates/livestreamStart.html');
        let emailHtml = fs.readFileSync(templatePath, 'utf8');
        
        emailHtml = emailHtml.replace('{{title}}', livestreamData.title);
        emailHtml = emailHtml.replace('{{description}}', livestreamData.description 
          ? `<p style="color: #666666; font-size: 16px; line-height: 1.6; margin: 0 0 30px 0;">${livestreamData.description}</p>` 
          : '');
        emailHtml = emailHtml.replace('{{watchUrl}}', watchUrl);
        
        for (const user of users.rows) {
          try {
            await sendEmail(user.email, 'Livestream Started', emailHtml);
          } catch (error) {
            console.error(`Failed to send email to ${user.email}:`, error);
          }
        }
      });
    }

    return { success: true, notificationsSent: notifications.length };
  } catch (error) {
    console.error('Notification service error:', error);
    throw error;
  }
};

File: backend/app/services/notification_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from app.models.notification import Notification
from app.models.user import User
from app.services.email_service import send_email

async def create_notification(
    db: AsyncSession,
    type: str,
    message: str,
    recipient_id: str = None,
    recipient_email: str = None,
    link: str = None
):
    """Create a notification"""
    notification = Notification(
        type=type,
        message=message,
        recipient_id=recipient_id,
        recipient_email=recipient_email,
        link=link,
        status="pending"
    )
    db.add(notification)
    await db.commit()
    await db.refresh(notification)
    return notification

async def send_notification_email(
    db: AsyncSession,
    notification_id: str,
    subject: str,
    html: str
):
    """Send notification via email"""
    result = await db.execute(
        select(Notification).where(Notification.id == notification_id)
    )
    notification = result.scalar_one_or_none()
    
    if notification and notification.recipient_email:
        try:
            await send_email(notification.recipient_email, subject, html)
            notification.status = "sent"
        except Exception as e:
            notification.status = "failed"
        
        await db.commit()
